### PR TITLE
Flag non-standard dither and apply monitor window checks as needed

### PR DESCRIPTION
**Pull request by jeanconn**
Obs 17383 in FEB2315A has a non-standard dither with a 64 arcsec amplitude in Z.

From Eric:

In order to have these dither parameters, dither commanding must be similar to that for observations with monitor windows.
 
Dither (using standard parameters) must be disabled 1 min prior to the end of the maneuver to the observation that will use the large dither parameters.  5 min after the end of the maneuver (when all stars should have been acquired and are being tracked), dither (using the large parameters) can be re-enabled.  Finally, 5 min prior to the end of the observation, dither must be commanded back to standard parameters.

### CODE REVIEW

**Comment by jeanconn** [comment]
Code attached.

**Comment by taldcroft** [comment]
I guess we rely on the new regression weeks for functional validation, though the coverage is not 100% (e.g. I don't think there is an example of the dither period being unexpected).

Hopper is coming along nicely and will support actual functional tests by allowing small backstop snippets to be run, including straightforward injection of mistakes like a wrong dither period.  In the very near future we should be thinking hopper for all new starcheck development.

**Comment by jeanconn** [comment]
We still need to get a dither history file that has amplitude and period.

**Comment by taldcroft** [comment]
We could always start down the path of relying on command states where History data is not available.  We already do this for thermal checking, though that is a bit squishier.

**Comment by taldcroft** [comment]
(That of course also requires putting this into command states, but it will be a snap with dynamic cmd states).

**Comment by jeanconn** [comment]
Certainly doable from cmd_states but this seems like a trivial initial
condition that should just be in a new dither history file.  SAUSAGE needs
it too, and I think we started to ask for it 4 years ago.​

**Comment by taldcroft** [comment]
FYI here is the implementation and testing for dither state handling in hopper (with a little bit of extra cruft thrown in):
https://github.com/sot/hopper/commit/c2a1a65781491efdb5d19731d8d4601badf4e5dc

Notice how the testing can be really specific and localized.

**Comment by jeanconn** [comment]
Is you implementation idea that we'll continue to develop hopper tests and use them in Perl starcheck, or develop on the side until you do a full switch to hopper?

**Comment by taldcroft** [comment]
BTW I improved and extended the implementation in hopper: https://github.com/sot/hopper/commit/afa397a4b22c475817ecb1c94084e58c7ca83f7d.

I would say that going forward we should be thinking of Python and hopper for any new starcheck tests, where this makes sense.  Basically we don't want to duplicate code.  Currently the API in hopper is stabilizing but there are likely to be tweaks as things go forward. In fact actually using hopper for starcheck tests will be a good way to identify weaknesses or issues with the API.

In this case this PR with perl code should be used (after review is finished).

**Comment by jeanconn** [comment]
Higher up in the conversation I had that question about AOACRSET.  Eric signed off on products without that command, so a check for it is not included in the extra "large dither" checks in this PR.

File: src/lib/Ska/Starcheck/Obsid.pm [hunk 1]
```diff
@@ -682,6 +713,63 @@ sub check_dither {
...
+    my $t_manv = $manv->{tstop};
+    my %dt = (AODSDITH => -60, AOENDITH => 300);
+    my %cnt = map { $_ => 0 } keys %dt;
+    my $all_dither_count = 0;
+    foreach my $bs (grep { $_->{cmd} eq 'COMMAND_SW' } @{$backstop}) {
+       my %param = Ska::Parse_CM_File::parse_params($bs->{params});
+       next unless $bs->{time} >= ($t_manv - ($dt{AODSDITH} + $time_tol + 1));
+       next unless ($param{TLMSID} =~ /^AO/);
+       # Search for all relevant dither commands
+       foreach my $cmd ('AODSDITH', 'AOENDITH', 'AODITPAR') {
```
Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: There is no check here for a return to standard parameters, so I think AODITPAR is a bit of a decoy.

File: src/lib/Ska/Starcheck/Obsid.pm [hunk 2]
```diff
@@ -619,9 +619,16 @@ sub check_dither {
     my $self = shift;
 
     my $dthr = shift;		  # Ref to array of hashes containing dither states
+    my $backstop = shift;
     my %dthr_cmd = (ON => 'ENAB',   # Translation from OR terminology to dither state term.
 		    OFF => 'DISA');
 
+    my %standard_y_dither = (20 => 1087.0,
```
Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: Need to include the other dither cases as well, probably encoded in a table like the one Erik provided.  That's my plan for hopper anyway.
```
Dither Case        Y Amplitude  Z Amplitude  Y Period  Z Period
-----------------  -----------  -----------  --------  --------
Standard ACIS           8.0          8.0      1000.0     707.1
Standard HRC           20.0         20.0      1087.0     768.6
Crab Observations       1.0          1.0      1000.0     707.1
ACIS Subarray           8.0          4.0      1000.0     707.4
ACIS Contaminant*       8.0         64.0      1000.0    2647.0
ACA DCC DFC             0.0          0.0      1000.0     707.1
```

Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: I was under the impression/of the opinion that flagging all but the first two as  "non-standard" would be fine for this PR and that the rest would belong more in #96

Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: I see your point.  Sounds like a good question for SSAWG.

Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: Sounded like no real objections from SSAWG on this approach.

File: src/lib/Ska/Starcheck/Obsid.pm (hunk 1 above)
Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: Add a check for a return to standard parameters.

Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: As mentioned today, from a system-level perspective the ACA checker should be responsible for making sure the dither parameters match the OR list, while the PCAD checker ensures that commanding is correct and safe for the commanded dither parameters.  Which is a way of saying you don't *have* to check for the return to standard parameters (and can spend effort elsewhere).

Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: But, if we have already checked 2/3 of the things that we don't need to check, I'd be reassured if we just checked this too.

Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: The "5 minutes before next maneuver" part of the requirement makes me think this won't be so easy to implement and will have corner case code.  If you are sure you can complete and fully test this new code in ~15 minutes then go for it, but I don't think there is any value added.

Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: As for "reassured", I have 100% confidence that Larry will do the backstop check correctly.

Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: My notes had 5 minutes before the end of the observation, which I was planning to just do via the or_er_tstop that should already be hanging around in the data structure for each obsid.

**Comment by jeanconn** [comment]
I'm also of course fine with chucking the disable/enable checks; I was just still working from your comment that  "I think we do always want these checks for the existing non-standard dither configurations,".

File: src/lib/Ska/Starcheck/Obsid.pm (hunk 1 above)
Comment (taldcroft) [inline src/lib/Ska/Starcheck/Obsid.pm]: Eric's notes say "- Return to standard ACIS parameters 5 min before start of next maneuver ...".  Not sure when those might be different.

Comment (jeanconn) [inline src/lib/Ska/Starcheck/Obsid.pm]: Gotcha.  I think we have two sets of Eric notes, because you opened this issue with the "Finally, 5 min prior to the end of the observation, dither must be commanded back to standard parameters."  But I think those are basically only different in the last observation in products.  For the last observation, I just use the "processing summary" stop time for the other checks (dither, momentum dumps).

**Comment by jeanconn** [comment]
Code attached to make the large dither check at end of observation more clear.  This based on the comments that ended up in #131.  Does this help @taldcroft ?

**Comment by jeanconn** [comment]
Now that you've given a :+1: , may I make one more change?  I had added a products that fail the large dither checks, but it would be good to add the B products too.

**Comment by taldcroft** [comment]
ok.

**Comment by jeanconn** [comment]
A few more updates responding to diffs in regression outputs.

**Comment by taldcroft** [comment]
:+1: on new commits.
